packages/graalvm/src/main/kotlin/elide/runtime/node/querystring/NodeQuerystring.kt

```diff
@@ -12,36 +12,356 @@
+        // Handle host objects directly without QueryParams conversion
+        val hostObject = obj.asHostObject<Any>()
+        if (hostObject is Map<*, *>) {
+          hostObject.forEach { (key, value) ->
+            if (key is String && value != null) {
+              appendKeyValuePairs(result, key, Value.asValue(value), equals, stringifyOptions)
+            }
+          }
+        } else {
+          // Non-map host objects should return empty string
+          return ""
+        }
+      }
+
+      else -> {
+        // Use QueryParams.fromGuest for guest objects
+        val queryParams = QueryParams.fromGuest(obj)
+        if (queryParams != null) {
+          queryParams.memberKeys.forEach { key ->
+            val value = queryParams.getMember(key)
+            appendKeyValuePairs(result, key, value, equals, stringifyOptions)
+          }
+        } else {
+          // Node.js returns empty string for non-objects (primitives, null, etc.)
+          return ""
+        }
+      }
+    }
+
+    return result.joinToString(separator)
+  }
+
+  private fun encodeValue(value: Any?, encoder: Value?): String {
+    return when (value) {
+      null -> ""
+      else -> encoder?.execute(Value.asValue(value))?.asString() ?: escape(Value.asValue(value))
+    }
+  }
+
+  private fun appendKeyValuePairs(
+    result: MutableList<String>,
+    key: String,
+    value: Any?,
+    equals: String,
+    options: StringifyOptions
+  ) {
+    val encoder = options.encodeURIComponent
+    val encodedKey = encoder?.execute(Value.asValue(key))?.asString() ?: escape(Value.asValue(key))
+
+    when (value) {
+      is ReadOnlyArrayProxy -> {
+        for (index in 0 until value.size) {
+          val encodedValue = encodeValue(value.get(index), encoder)
+          result += "$encodedKey$equals$encodedValue"
+        }
+      }
+
+      is Value -> {
+        when {
+          value.isNull -> result += "null"
+          value.isHostObject -> {
+            val hostObject = value.asHostObject<Any>()
+            if (hostObject is Array<*>) {
+              hostObject.forEach { element ->
+                val encodedValue = encodeValue(element, encoder)
+                result += "$encodedKey$equals$encodedValue"
+              }
+            }
+          }
+          else -> {
+            val encodedValue = encodeValue(valueToString(value), encoder)
+            result += "$encodedKey$equals$encodedValue"
+          }
+        }
+      }
+
+      else -> {
+        val encodedValue = encodeValue(value.toString(), encoder)
+        result += "$encodedKey$equals$encodedValue"
+      }
+    }
+  }
+
+  override fun unescape(str: Value): String {
+    val input = valueToString(str)
+
+    return runCatching {
+      val bytes = input.toByteArray(StandardCharsets.UTF_8)
+      codec.decode(bytes).toString(Charsets.UTF_8)
+    }.getOrElse {
+      // Fallback: manually decode percent-encoded sequences
+      input.replace(Regex("%([0-9A-Fa-f]{2})")) { match ->
```

Comment:
Regular expressions should be stored as static members in the companion to avoid creating and compiling the pattern on each function call (and also for easy access)